appendix_abstractions.asciidoc

```diff
@@ -282,23 +450,27 @@ def test_when_a_file_has_been_renamed_in_the_source():
     synchronise_dirs(reader.pop, actions.append, "/source", "/dest")
 
     assert actions = [("MOVE", "/dest/original-file", "/dest/renamed-file")]
-
 ----
 ====
 
-Because we've disentangled the logic of our program - the code for identifying
-changes - from the low-level details of IO, we can easily test the code. All we
-need to do is plug in two functions - one for generating hashes and file names,
-and the other for executing the resulting actions list.
+The advantage of this approach is that your tests act on the exact same function 
+that's used by your production code.  The disadvantage is that DI usually demands
+a bit more work on the part of the reader to understand what's going on.
+
+TODO: discuss the fact that our implementation is full of bugs and now we
+    can easily test for them.
+
+
+=== Wrap-up: "Depend on Abstractions".
 
-This easier testing isn't the only benefit here, though. Consider a new feature
-requirement: we would like to add a "dry-run" mode to our program. If a flag is
-passed on the command line, then our program should output the changes it will
-make, but not actually execute any of them.
+Choosing the right abstraction is very much an art. 
 
-Which of the two versions is easier to extend in this way?
+* What's the simplest data structure I can use to represent these dependencies?
+* Where can I draw a line between my systems, where can I carve out a seam, to
+  stick that abstraction in?
+* What are the dependencies and what is the core business logic anyway?
 
-TODO: continue example to make the point.
+Practice makes less-imperfect!
 
-TODO: investigate using `pathlib` to make it all a bit easier to read?
+TODO: can we do a better wrap-up?
```

Comment:
"Wel'll see this idea come up again and again in the book: we can make our systems easier to test and maintain by simplifying the interface between our business logic and messy IO. Finding the right abstraction is tricky, but one useful heuristic is to choose a familiar Python datastructure that can represent the state of the messy system, and try to imagine a single function that can return that state."

That's off the cuff and trails off into uselessness, but is that the right kinda thing?

Reply:
yep! i'll add that for now and leave the TODO maybe?

Reply:
now that i'm trying to add it in, it is a bit repeatey of what we already have...

Reply:
ok i got it in